plugins/usbdmx/JaRuleWidgetPort.h

```diff
@@ -0,0 +1,202 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * JaRuleWidgetPort.h
+ * A Ja Rule widget port.
+ * Copyright (C) 2015 Simon Newton
+ */
+
+#ifndef PLUGINS_USBDMX_JARULEWIDGETPORT_H_
+#define PLUGINS_USBDMX_JARULEWIDGETPORT_H_
+
+#include <libusb.h>
+#include <ola/Callback.h>
+#include <ola/io/ByteString.h>
+#include <ola/thread/ExecutorInterface.h>
+#include <ola/thread/Mutex.h>
+#include <ola/util/SequenceNumber.h>
+
+#include <map>
+#include <queue>
+
+#include "plugins/usbdmx/JaRulePortHandle.h"
+#include "plugins/usbdmx/LibUsbAdaptor.h"
+#include "plugins/usbdmx/JaRuleWidget.h"
+
+namespace ola {
+namespace plugin {
+namespace usbdmx {
+namespace jarule {
+
+/**
+ * @brief The internal model of a port on a JaRule device.
+ *
+ * Each port has its own libusb transfers as well as a command queue. This
+ * avoids slow commands on one port blocking another.
+ */
+class JaRuleWidgetPort {
+ public:
+  /**
+   * @brief Create a new JaRuleWidgetPort.
+   * @param executor The Executor to run the callbacks on.
+   * @param adaptor The LibUsbAdaptor to use.
+   * @param usb_handle the libusb_device_handle for the Ja Rule widget.
+   * @param endpoint_number The endpoint to use for transfers.
+   * @param uid The device's UID.
+   * @param physical_port The physical port index.
+   */
+  JaRuleWidgetPort(ola::thread::ExecutorInterface *executor,
+                   ola::plugin::usbdmx::LibUsbAdaptor *adaptor,
+                   libusb_device_handle *usb_handle,
+                   uint8_t endpoint_number,
+                   const ola::rdm::UID &uid,
+                   uint8_t physical_port);
+
+  /**
+   * @brief Destructor
+   */
+  ~JaRuleWidgetPort();
+
+  /**
+   * @brief Claim the handle to this port.
+   * @returns a port handle, ownership is not transferred. Will return NULL if
+   *   the port is already claimed.
+   */
+  JaRulePortHandle* ClaimPort();
+
+  /**
+   * @brief Release a handle to a port.
+   * @returns a port handle, ownership is not transferred.
+   */
+  void ReleasePort();
+
+  /**
+   * @brief Cancel all commands for this port.
+   */
+  void CancelAll();
+
+  /**
+   * @brief Send a command on this port.
+   * @param command the Command type.
+   * @param data the payload data. The data is copied and can be freed once the
+   *   method returns.
+   * @param size the payload size.
+   * @param callback The callback to run when the message operation completes.
+   * This may be run immediately in some conditions.
+   *
+   * SendMessage can be called from any thread, and messages will be queued.
```

Comment:
No inheritance here, ones public facing and the other is the internal implementation.